packages/opentelemetry-types/src/trace/span.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SpanContext } from './span_context';
+import { Status } from './status';
+
+/**
+ * An interface that represents a span. A span represents a single operation
+ * within a trace. Examples of span might include remote procedure calls or a
+ * in-process function calls to sub-components. A Trace has a single, top-level
+ * "root" Span that in turn may have zero or more child Spans, which in turn
+ * may have children.
+ */
+export interface Span {
+  /**
+   * Returns the SpanContext object associated with this Span.
+   *
+   * @returns the SpanContext object associated with this Span.
+   */
+  context(): SpanContext;
+
+  // /**
+  //  * # TODO
+  //  * Returns the Tracer object used to create this Span.
+  //  * https://github.com/open-telemetry/opentelemetry-specification/issues/21
+  //  */
+  // tracer(): Tracer;
+
+  /**
+   * Sets an attribute to the span.
+   *
+   * @param key the key for this attribute.
+   * @param value the value for this attribute. If the value is a typeof object
+   *     it has to be JSON.stringify-able otherwise it will raise an exception.
+   */
+  setAttribute(key: string, value: string | number | boolean | object): void;
+
+  /**
+   * Adds an event to the Span.
+   *
+   * @param name the name of the event.
+   * @param [attributes] the attributes that will be added; these are
+   *     associated with this event.
+   */
+  addEvent(
+    name: string,
+    attributes?: { [key: string]: string | number | boolean | object }
```

Comment:
Similar to the above, `attributes` should be of type `any`.

Reply:
I am fine with changing it to `any` type, but will lose type-safety and to mitigate this I have to add TSLint rule flag (`tslint:disable-next-line:no-any`). WDYT?

Reply:
Since there is a limited number of types in JS it would be possible to simply specify all of them, but I think `any` is clearer that the intent is to accept anything. I'm personally fine with a TSLint comment. If it happens to much we could simply remove the rule. Code reviews will ensure that there is no abuse.

Reply:
Ok, done.

Reply:
Could we use `unknown` instead of `any`? That allows the same semantics (the passed in value can be any type), but it requires type guards or a cast to actually do something with the type and so is safer from a type checking perspective. As a plus it doesn't require the lint comment. (Most times where you say `any` you can usually say `unknown` and get the same effect)

Reply:
Would it make more sense then to include every possible JavaScript types? There are not that many and it would remove the need for casting for common types.

Reply:
> Could we use unknown instead of any?

For now, I have changed to `unknown `. If needed, will change to either `any` or every possible JavaScript types.

Reply:
My understanding is that `unknown` is basically equivalent (maybe exactly equivalent?) to the union type of every possible type. You can use `typeof` and `instanceof` that would do runtime checks but are also used by the TS compiler to infer a more narrow type at compile time for code in code blocks guarded by those checks. Casts would only be needed if you somehow know what type it is and force it to that, which perhaps something vendor-specific might do.

See the `function f20(x: unknown)` example in the  in the [TS 3.0 release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html) for an example of how `typeof` and `instanceof` can be used for type narrowing by the compiler (and of course checking of the value's actual JS runtime type so it can be correctly serialized).